### PR TITLE
[Enhancement] try to buffer aggregated data in spillable agg operator

### PR DESCRIPTION
**Pull request by silverbullet233**
## background

If the SpillableAggregateBlockingSinkOperator triggers spilling, each chunk will be aggregated once in the hash table, and then immediately added to the spiller. Once data added into the MemTable of Spiller, it won't be further aggregated. 

Considering that the MemTable of Spiller also buffers data in memory before flushing, here is an obvious optimization strategy. We can allocate more memory to the hash table and use it to aggregate data to reduce the amount of data written to disk.

## optimization idea

I added a pre-aggregated data strategy in this PR, the basic idea is similar to the AggregateStreamingSinkOperator.

We still maintain a hash table in memory, and use it to aggregate data. 
* If the aggregation degree of the data is high, we will try to keep the hash table to aggregate more data. 
* If the aggregation degree is low which means that there is little point in caching the hash table in memory, so just flush  it to disk and then rebuild.

Different from agg streaming, the purpose of triggering spilling is to reduce memory usage, here we will strictly control the size of the hash table in memory.

## test result

* data set: tpc-ds 1T
* environment: 1 FE + 1 BE, both are deployed on an instance with 16c and 64g.
I changed some session variables so that some complex query can enter spill mode.
```sql
set query_mem_limit=26843545600;
set spill_mem_limit_threshold=0.6;
set spill_operator_min_bytes=8388608;
```
In addition, I compared the optimization effect by controlling the environment variable enable_agg_spill_preaggregation.
when enable_agg_spill_preaggregation=false, agg sink operator won't preaggregate data, just as before.
when enable_agg_spill_preaggregation=true, this optimization is used.

I test all queries of tpcds and the following queries can benefit from this optimization.

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"><head></head><body>

  | Agg Node Count in Plan | QueryTime-baseline | QueryTime-opt | QueryTime-opt/baseline | SpillBytes-baseline | SpillBytes-opt | SpillBytes-baseline/opt | PeakMemUsage-baseline | PeakMemUsage-opt
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
Q04 | 6 | 32m40s | 20m56s | 0.64 | 100.866 GB | 63.628 GB | 0.63 | 21.237 GB | 20.657 GB
Q11 | 4 | 15m4s | 13m35s | 0.90 | 58.886 GB | 51.706 GB | 0.87 | 15.367 GB | 15.381 GB
Q23-1 | 3 | 16m28s | 10m9s | 0.62 | 54.179 GB | 30.602 GB | 0.56 | 16.190 GB | 16.672 GB
Q23-2 | 5 | 16m32s | 10m5s | 0.60 | 55.269 GB | 28.812 GB | 0.52 | 16.244 GB | 15.712 GB
Q67 | 1 | 27m19s | 16m22s | 0.60 | 49.683 GB | 26.238 GB | 0.52 | 15.000 GB | 14.996 GB

</body></html>

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [null Incremental Coverage Report]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|6|100.00%|[]|

File: be/src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.cpp [hunk 1]
```diff
@@ -95,41 +103,165 @@ Status SpillableAggregateBlockingSinkOperator::prepare(RuntimeState* state) {
...
+}
+
+void SpillableAggregateBlockingSinkOperator::_add_non_agg_chunk(ChunkPtr chunk) {
+    _non_agg_rows += chunk->num_rows();
+    _non_agg_bytes += chunk->memory_usage();
+
+    _non_agg_chunks.push(std::move(chunk));
+}
+
+Status SpillableAggregateBlockingSinkOperator::_try_to_spill_by_auto(RuntimeState* state, const ChunkPtr& chunk) {
```
Comment (fzhedu) [inline be/src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.cpp:145]: can it consider with auto agg? use such window strage do decide pash through, preagg, spill?

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 11 / 77 (14.29%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.h|0|4|00.00%|[47, 73, 78, 80]|
|:large_blue_circle:|src/exec/aggregator.cpp|0|2|00.00%|[511, 512]|
|:large_blue_circle:|src/exec/spill/spiller.hpp|0|2|00.00%|[81, 82]|
|:large_blue_circle:|src/exec/pipeline/hashjoin/hash_join_build_operator.cpp|0|1|00.00%|[54]|
|:large_blue_circle:|src/runtime/runtime_state.h|0|5|00.00%|[332, 336, 337, 338, 340]|
|:large_blue_circle:|src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.cpp|0|42|00.00%|[48, 49, 52, 59, 62, 64, 65, 66, 67, 68, 69, 75, 115, 116, 117, 125, 126, 127, 128, 130, 132, 133, 134, 149, 151, 152, 153, 154, 155, 156, 157, 159, 162, 164, 166, 167, 169, 171, 173, 174, 175, 177]|
|:large_blue_circle:|src/exec/pipeline/exchange/exchange_sink_operator.cpp|0|4|00.00%|[155, 242, 445, 446]|
|:large_blue_circle:|src/exec/pipeline/aggregate/spillable_aggregate_blocking_source_operator.cpp|0|1|00.00%|[36]|
|:large_blue_circle:|src/runtime/local_pass_through_buffer.cpp|1|6|16.67%|[27, 91, 92, 96, 123]|
|:large_blue_circle:|src/exec/spill/spiller.h|2|2|100.00%|[]|
|:large_blue_circle:|src/exec/spill/input_stream.cpp|2|2|100.00%|[]|
|:large_blue_circle:|src/exec/spill/spiller.cpp|3|3|100.00%|[]|
|:large_blue_circle:|src/exec/spill/mem_table.cpp|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 13 / 83 (15.66%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/pipeline/aggregate/aggregate_streaming_source_operator.cpp|0|1|00.00%|[88]|
|:large_blue_circle:|src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.h|0|4|00.00%|[73, 77, 78, 80]|
|:large_blue_circle:|src/runtime/local_pass_through_buffer.cpp|0|5|00.00%|[27, 91, 92, 96, 136]|
|:large_blue_circle:|src/exec/spill/spiller.hpp|0|2|00.00%|[81, 82]|
|:large_blue_circle:|src/exec/pipeline/hashjoin/hash_join_build_operator.cpp|0|1|00.00%|[54]|
|:large_blue_circle:|src/exec/pipeline/aggregate/aggregate_distinct_streaming_source_operator.cpp|0|1|00.00%|[79]|
|:large_blue_circle:|src/runtime/runtime_state.h|0|5|00.00%|[334, 336, 337, 338, 340]|
|:large_blue_circle:|src/exec/pipeline/aggregate/spillable_aggregate_distinct_blocking_operator.h|0|1|00.00%|[54]|
|:large_blue_circle:|src/exec/pipeline/pipeline_driver_executor.cpp|0|2|00.00%|[171, 172]|
|:large_blue_circle:|src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.cpp|0|41|00.00%|[52, 57, 59, 60, 61, 62, 106, 107, 108, 117, 119, 120, 123, 124, 125, 126, 127, 141, 142, 143, 145, 146, 149, 151, 152, 153, 154, 155, 156, 157, 159, 162, 164, 166, 167, 169, 171, 173, 174, 175, 177]|
|:large_blue_circle:|src/exec/pipeline/exchange/exchange_sink_operator.cpp|0|4|00.00%|[155, 242, 445, 446]|
|:large_blue_circle:|src/exec/pipeline/aggregate/spillable_aggregate_blocking_source_operator.cpp|0|1|00.00%|[36]|
|:large_blue_circle:|src/exec/aggregator.cpp|4|6|66.67%|[490, 904]|
|:large_blue_circle:|src/exec/spill/spiller.h|2|2|100.00%|[]|
|:large_blue_circle:|src/exec/stream/aggregate/stream_aggregator.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/exec/spill/spiller.cpp|3|3|100.00%|[]|
|:large_blue_circle:|src/exec/spill/mem_table.cpp|3|3|100.00%|[]|

File: be/src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.cpp (hunk 1 above)
Comment (silverbullet233) [reply be/src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.cpp:151]: > can it consider with auto agg? use such window strage do decide pash through, preagg, spill?

The current strategy is to learn from streaming auto agg

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 3 / 110 (02.73%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/hash_joiner.cpp|0|1|00.00%|[57]|
|:large_blue_circle:|src/exec/spill/spiller.hpp|0|1|00.00%|[80]|
|:large_blue_circle:|src/exec/pipeline/hashjoin/hash_join_build_operator.cpp|0|1|00.00%|[54]|
|:large_blue_circle:|src/runtime/runtime_state.h|0|4|00.00%|[335, 337, 338, 339]|
|:large_blue_circle:|src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.cpp|0|98|00.00%|[56, 57, 58, 59, 105, 107, 108, 117, 118, 119, 120, 123, 124, 126, 139, 140, 141, 142, 145, 146, 147, 148, 151, 152, 153, 155, 156, 157, 159, 160, 165, 168, 169, 170, 171, 172, 173, 175, 176, 177, 178, 180, 181, 182, 186, 187, 190, 192, 194, 195, 196, 197, 199, 200, 204, 205, 208, 209, 210, 211, 214, 215, 218, 219, 223, 224, 225, 228, 229, 230, 231, 234, 237, 240, 241, 242, 243, 246, 250, 252, 253, 254, 255, 258, 259, 260, 261, 262, 264, 266, 267, 268, 269, 271, 272, 292, 293, 295]|
|:large_blue_circle:|src/exec/spill/spill_components.cpp|1|2|50.00%|[475]|
|:large_blue_circle:|src/exec/aggregator.cpp|2|3|66.67%|[904]|

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 2]
```diff
@@ -845,6 +848,15 @@ public static MaterializedViewRewriteMode parse(String str) {
...
     private int spillEncodeLevel = 7;
 
+    @VarAttr(name = ENABLE_AGG_SPILL_PREAGGREGATION, flag = VariableMgr.INVISIBLE)
+    public boolean enableAggSpillPreaggregation = true;
+
+    @VarAttr(name = AGG_SPILL_HT_LOW_REUCTION_THRESHOLD, flag = VariableMgr.INVISIBLE)
+    public double aggSpillHtLowReductionThreshold = 0.5;
+
+    @VarAttr(name = AGG_SPILL_HT_LOW_REDUCTION_CHUNK_LIMIT, flag = VariableMgr.INVISIBLE)
+    public int aggSpillhtLowReductionChunkLimit = 5;
```
Comment (fzhedu) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:858]: need adjust these ? if yes, use force is easier?

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|6|6|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 5 / 112 (04.46%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/hash_joiner.cpp|0|1|00.00%|[57]|
|:large_blue_circle:|src/exec/spill/spiller.hpp|0|1|00.00%|[80]|
|:large_blue_circle:|src/exec/pipeline/hashjoin/hash_join_build_operator.cpp|0|1|00.00%|[54]|
|:large_blue_circle:|src/runtime/runtime_state.h|0|4|00.00%|[335, 337, 338, 339]|
|:large_blue_circle:|src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.cpp|0|98|00.00%|[56, 57, 58, 59, 105, 107, 108, 117, 118, 119, 120, 123, 124, 126, 139, 140, 141, 142, 145, 146, 147, 148, 151, 152, 153, 155, 156, 157, 159, 160, 165, 168, 169, 170, 171, 172, 173, 175, 176, 177, 178, 180, 181, 182, 186, 187, 190, 192, 194, 195, 196, 197, 199, 200, 204, 205, 208, 209, 210, 211, 214, 215, 218, 219, 223, 224, 225, 228, 229, 230, 231, 234, 237, 240, 241, 242, 243, 246, 250, 252, 253, 254, 255, 258, 259, 260, 261, 262, 264, 266, 267, 268, 269, 271, 272, 292, 293, 295]|
|:large_blue_circle:|src/exec/spill/spill_components.cpp|1|2|50.00%|[475]|
|:large_blue_circle:|src/exec/aggregator.cpp|2|3|66.67%|[904]|
|:large_blue_circle:|src/exec/spill/spiller.cpp|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 2 above)
Comment (silverbullet233) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:858]: > need adjust these ? if yes, use force is easier?

maybe we don't need expose these configurations to user, I will remove them

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 5 / 107 (04.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/hash_joiner.cpp|0|1|00.00%|[57]|
|:large_blue_circle:|src/exec/spill/spiller.hpp|0|1|00.00%|[80]|
|:large_blue_circle:|src/exec/pipeline/hashjoin/hash_join_build_operator.cpp|0|1|00.00%|[54]|
|:large_blue_circle:|src/runtime/runtime_state.h|0|1|00.00%|[337]|
|:large_blue_circle:|src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.cpp|0|96|00.00%|[56, 57, 58, 59, 105, 115, 116, 117, 118, 121, 122, 124, 137, 138, 139, 140, 143, 144, 145, 146, 149, 150, 151, 153, 154, 155, 157, 158, 163, 166, 167, 168, 169, 170, 171, 173, 174, 175, 176, 178, 179, 180, 184, 185, 188, 190, 192, 193, 194, 195, 197, 198, 202, 203, 206, 207, 208, 209, 212, 213, 216, 217, 221, 222, 223, 226, 227, 228, 229, 232, 235, 238, 239, 240, 241, 244, 248, 250, 251, 252, 253, 256, 257, 258, 259, 260, 262, 264, 265, 266, 267, 269, 270, 290, 291, 293]|
|:large_blue_circle:|src/exec/spill/spill_components.cpp|1|2|50.00%|[467]|
|:large_blue_circle:|src/exec/aggregator.cpp|2|3|66.67%|[904]|
|:large_blue_circle:|src/exec/spill/spiller.cpp|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 5 / 107 (04.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/hash_joiner.cpp|0|1|00.00%|[57]|
|:large_blue_circle:|src/exec/spill/spiller.hpp|0|1|00.00%|[80]|
|:large_blue_circle:|src/exec/pipeline/hashjoin/hash_join_build_operator.cpp|0|1|00.00%|[54]|
|:large_blue_circle:|src/runtime/runtime_state.h|0|1|00.00%|[337]|
|:large_blue_circle:|src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.cpp|0|96|00.00%|[56, 57, 58, 59, 105, 115, 116, 117, 118, 121, 122, 124, 137, 138, 139, 140, 143, 144, 145, 146, 149, 150, 151, 153, 154, 155, 157, 158, 163, 166, 167, 168, 169, 170, 171, 173, 174, 175, 176, 178, 179, 180, 184, 185, 188, 190, 192, 193, 194, 195, 197, 198, 202, 203, 206, 207, 208, 209, 212, 213, 216, 217, 221, 222, 223, 226, 227, 228, 229, 232, 235, 238, 239, 240, 241, 244, 248, 250, 251, 252, 253, 256, 257, 258, 259, 260, 262, 264, 265, 266, 267, 269, 270, 290, 291, 293]|
|:large_blue_circle:|src/exec/spill/spill_components.cpp|1|2|50.00%|[467]|
|:large_blue_circle:|src/exec/aggregator.cpp|2|3|66.67%|[904]|
|:large_blue_circle:|src/exec/spill/spiller.cpp|2|2|100.00%|[]|

**Review by fzhedu — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**